carbon-design-system/carbon
# refactor(styles,react): update layer styles to work in inline themes

Closes #10049 

This component updates our layering system to work in inline themes (aka zones). Currently, our system is flawed because it relies on the nesting of selectors. In other words, we expect the nesting of the `.cds--layer` class to determine its level. This does not work in inline themes because these areas of the page "reset" their layer level.

More specifically, we would expect the following to work for the `<Layer>` component:

```jsx
<div>By default, use an 01 layer token for a `layer` token</div>
<Layer>
  <div>Use an 02 layer token for a `layer` token</div>
  <Layer>
    <div>Use an 03 layer token for a `layer` token</div>
  </Layer>
</Layer>
```

This works in our current implementation, and each `Layer` component is essentially a `<div>` with the `cds--layer` class. Each level of the `cds--layer` class maps the `layer` token to the appropriate level:

```css
// We start at two since `--cds-layer` maps to `--cds-layer-01` by default
.cds--layer {
  --cds-layer: var(--cds-layer-02);
}

.cds--layer .cds--layer {
  --cds-layer: var(--cds-layer-03);
}
```

However, this does _not_ work in our inline themes as the nesting rules no longer apply.

```jsx
<div>By default, use an 01 layer token for a `layer` token</div>
<Layer>
  <div>Use an 02 layer token for a `layer` token</div>
  <Theme theme="g100">
    <div>Use an 01 layer token for a `layer` token</div>
  </Theme>
</Layer>
```

Here, the inline theme _does not_ use an 01 token and instead uses an 02 token value.

This PR addresses this problem by creating dedicated classes for layer sets:

- `.cds--layer-one`
- `.cds--layer-two`
- `.cds--layer-three`

We now use context to communicate the "layer" inside of React. This will correctly apply the layer class and allows us to "reset" the layer when using an inline theme.


#### Changelog

**New**

- Add a new `_layer.scss` entrypoint

**Changed**

- Update our styles entrypoint and `_zone.scss` entrypoint
- Update our Theme component to correct set the initial layer class
- Update our storybook to correctly pick-up side effects from our files
- Update our storybook to correctly set the default theme

**Removed**

- Remove `_layer-set.scss`

#### Testing / Reviewing

- View the layer storybook stories and verify they work as intended
- View the theme "with layer" story and verify it works as intended
